---
title: "Introduction to gep2mep"
author: "Francesco Napolitano"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## About gep2mep

Gene Module Expression Profiles (MEPs) are based on the expression of
gene modules (sets of genes) as opposed to individual genes. *gep2mep*
supports the convertion of gene expression profiles (GEPs) to MEPs and
can perform enrichment analysis of both modules and
perturbagens.

*gep2mep* creates a local repository of gene modules, which can also
be imported from the MSigDB database. The local repository is in the
*repo* format. When a GEP is passed to the function *buildMEPs*, it
refers to the stored database of modules to convert the GEP to a MEP
and permanently store the latter.

One type of analysis that can be performed on MEPs and that is
directly supported by *gep2mep* is the Drug-Set Enrichment Analysis
(DSEA, see reference below). It finds gene modules that are
consistently dysregulated by a set of drugs, as opposed to a
background of other drugs. Of course MEPs may refer to
non-pharmacological perturbagens (genetic perturbations, disease
states, etc.) for analogous analyses. See the *PertSEA* function

A complementary approach is that of finding perturbagens that
consistently dysregulate a set of gene modules. This is the
module-based version of the Gene Set Enrichment Analysis (GSEA). See
ModSEA.


## Creating a repository of gene modules.

The [MSigDB](http://software.broadinstitute.org/gsea/msigdb) is a
curated database of gene set collections. The entire database can be
downloaded as a single XML file and used by *gep2mep*. The following
commented code would import the database once downloaded:

```{r}
## db <- importMSigDB.xml("msigdb_v6.0.xml")
```

However, for this vignette a small excerpt will be used:

```{r}
db <- readRDS(system.file("testgmd.RDS", package="gep2mep"))
```

The database includes 30 gene modules, each of which is included in
one of 3 different collections. Each collection is identified by a
"db" and "subdb" name, which *gep2mep* puts together into a single
collection name using *buildDBids*.


```{r}
dbs <- sapply(db, get, x="db")
subdbs <- sapply(db, get, x="subdb")
print(dbs)
print(subdbs)
buildDBids(db)
```

In order to build a local *gep2mep* repository containing the gene
module data, *createRepository* is used:

```{r}
repoRoot <- file.path(tempdir(), "gep2mep_data")
rp <- createRepository(repoRoot, db)
```

The repository is in *repo* format, see the *repo* package help to use
the data directly. However, knowing *repo* is not necessary to use
*gep2mep*. The following lists the contents of the repository. Items
containing gene modules are assigned the name "DB_SUBDB_gmd".

```{r}
rp
rp$get("C3_TFT_gmd")$M5067$desc
```


## Adding Module Expression Profiles

Module Expression Profiles (MEPs) are created from Gene Expression
Profiles (GEPs) using gene module information from the
repository. GEPs must be provided as a matrix with rows corresponding
to genes and columns corresponding to conditions
(*perturbagens*). Genes and perturbagens must be specified through row
and column names respectively. The values must be ranks: for each
condition, the genes must be ranked from that being most UP-regulated
(rank 1) to that being most DOWN-regulated (rank equal to the number
of rows of the matrix).

One well known database that can be obtained in this format is for
example the Connectivty Map. A small excerpt (after further
processing) is included with the *gep2mep*. The excerpt must be
considered as a dummy example, as it only includes 500 genes for 5
perturbagens. It can be loaded as follows:

```{r}
geps <- readRDS(system.file("testgep.RDS", package="gep2mep"))
dim(geps)
geps[1:5, 1:3]
```

The GEPs can be converted to MEPs using the *buildMEPs* function. They
are stored as repository items by the names "DB_SUBDB". Each MEP is
composed of an Enrichment Score (ES) -- p-value (PV) pair associated
to each gene module. ESs and PVs are stored in two separated
matrices. For each perturbagen, the p-value reports wether a gene
module is significantly dysregulated and the sign of the corresponding
ES indicates the direction (UP- or DOWN-regulation).

```{r}
buildMEPs(rp, geps)
rp$get("C3_TFT")
```

## Performing Perturbagen-Set Enrichment Analysis.

Suppose the stored MEPs correspond to pharmacological
perturbations. Then *gep2mep* can perform Drug-Set Enrichment Analysis
(DSEA, see Napolitano et al., 2016, Bioinformatics). It finds gene
modules that are consistently dysregulated by a set of drugs, as
opposed to a background of other drugs. Of course MEPs may refer to
non-pharmacological conditions (genetic perturbations, disease states,
etc.) for analogous analyses (Perturbagen-Set Enrichment Analysis,
PertSEA). Given a set *dset* of drugs of interest, PertSEA (which in
this case is a DSEA) is performed as follows:

```{r}
pgset <- c("(+)_chelidonine", "(+/_)_catechin")
psea <- PertSEA(rp, pgset)
```

The result is a list of of 2 elements, named "PertSEA" and "details",
the most important of which is the former. The "PertSEA" entry is a
list in which each element refers to a gene module collection and is
made of a 2-columns matrix including ESs and p-values for each gene
module.

```{r}
psea$PertSEA$C3_TFT
```

In this dummy example the statistical background is made of only 3
GEPs (we added 5 in total), thus, as expected, there are no
significant p-values. For the C3_MIR collection, the pathway most
UP-regulated by the chosen set of two drugs is M5012, while the most
DOWN-regulated is M18759. They are respectively described as:

```{r}
gmds <- rp$get("C3_MIR_gmd")
gmds$M5012$desc
gmds$M18759$desc
```

The analysis can be exported in XLS format as follows:

```{r, eval=FALSE}
exportSEA(rp, psea)
```


## Performing Module-Set Enrichment Analysis

A complementary approach to PertSEA is Module-Set Enrichment Analysis
(ModSEA). ModSEA searches for perturbagens that consistently
dysregulate a set of gene modules. It can be seen as a module-based
version of the popular Gene Set Enrichment Analysis (GSEA). Since
modules are different in each collection, a different set must be
specified to run the analysis in each of them.

```{r}
modsets <- list(C3_TFT = c("M11607", "M10817", "M16694"),
                C4_CGN = c("M19723", "M5038", "M13419", "M1094"))
msea <- ModSEA(rp, modsets)
msea$C3_TFT
```

ModSEA results are analogous to those of PertSEA, but
perturbagen-wise.

```{r include=FALSE}
unlink(repoRoot, TRUE)
```