---
title: "Introduction to gep2pep"
date: 
author: "Francesco Napolitano"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Introduction to gep2pep}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## About gep2pep

Pathway Expression Profiles (PEPs) are based on the expression of
pathways (or generic gene sets) belonging to a collection, as opposed
to individual genes. `gep2pep` supports the conversion of gene
expression profiles (GEPs) to PEPs and performs enrichment analysis of
pathways or perturbagens.

`gep2pep` creates a local repository of gene sets, which can also be
imported from the MSigDB database [1]. The local repository is in the
`repo` format. When a GEP, defined as a ranked list of genes, is
passed to `buildPEPs`, the stored database of pathways is used to
convert the GEP to a PEP and permanently store the latter.

One type of analysis that can be performed on PEPs and that is
directly supported by `gep2pep` is the Drug-Set Enrichment Analysis
(DSEA, see reference below). It finds pathways that are consistently
dysregulated by a set of drugs, as opposed to a background of other
drugs. Of course PEPs may refer to non-pharmacological perturbagens
(genetic perturbations, disease states, etc.) for analogous
analyses. See the `PertSEA` function.

A complementary approach is that of finding perturbagens that
consistently dysregulate a set of pathways. This is the pathway-based
version of the Gene Set Enrichment Analysis (GSEA). As an application
example, this approach can be used to find drugs mimicking the
dysregulation of a gene by looking for drugs dysregulating pathways
involving the gene (this has been published as the `gene2drug` tool
[3]). See the `PathSEA` function.

This vignette uses toy data, as real data is computationally
expensive. Connectivity Map data [4] (drug induced gene expression
profiles) pre-converted to PEPs can be downloaded from
http://dsea.tigem.it in the `gep2pep` format.


### References

[1] Subramanian A. et al. Gene set enrichment analysis: A
    knowledge-based approach for interpreting genome-wide expression
    profiles. PNAS 102, 15545-15550 (2005).
    
[2] Napolitano F. et al, Drug-set enrichment analysis: a novel tool to
    investigate drug mode of action. Bioinformatics 32, 235-241 (2016).

[3] Napolitano F. et al, gene2drug: a Computational Tool for
    Pathway-based Rational Drug Repositioning, bioRxiv (2017) 192005;
    doi: https://doi.org/10.1101/192005

[4] Lamb, J. et al. The Connectivity Map: Using Gene-Expression
    Signatures to Connect Small Molecules, Genes, and Disease. Science
    313, 1929-1935 (2006).


## Creating a repository of pathways

In order to use the package, it must be loaded as follows:

```{r}
library(gep2pep)
```

The [MSigDB](http://software.broadinstitute.org/gsea/msigdb) is a
curated database of gene set collections. The entire database can be
downloaded as a single XML file and used by `gep2pep`. The following
commented code would import the database once downloaded:

```{r}
## db <- importMSigDB.xml("msigdb_v6.0.xml")
```

However, for this vignette a small excerpt will be used:

```{r}
db <- readRDS(system.file("testgmd.RDS", package="gep2pep"))
```

The database includes 30 pathways, each of which is included in one of
3 different collections. Each collection is identified by a "category"
and "subcategory" name, which `gep2pep` puts together into a single
collection identifier using `makeCollectionIDs`.


```{r}
cats <- sapply(db, get, x="category")
subcats <- sapply(db, get, x="subcategory")
print(cats)
print(subcats)
makeCollectionIDs(db)
```

In order to build a local `gep2pep` repository containing pathway
data, `createRepository` is used:

```{r}
repoRoot <- file.path(tempdir(), "gep2pep_data")
rp <- createRepository(repoRoot, db)
```

The repository is in `repo` format (see later in this document how to
access the repository directly). However, knowing `repo` is not
necessary to use `gep2pep`. The following lists the contents of the
repository and loads the description of the pathway "M5067" from the
collection "C3_TFT":

```{r}
rp
rp$get("C3_TFT_sets")$M5067$desc
```


## Creating Pathway Expression Profiles

Pathway Expression Profiles (PEPs) are created from Gene Expression
Profiles (GEPs) using pathway information from the
repository. GEPs must be provided as a matrix with rows corresponding
to genes and columns corresponding to conditions
(*perturbagens*). Genes and perturbagens must be specified through row
and column names respectively. The values must be ranks: for each
condition, the genes must be ranked from that being most UP-regulated
(rank 1) to that being most DOWN-regulated (rank equal to the number
of rows of the matrix).

One well known database that can be obtained in this format is for
example the Connectivty Map. A small excerpt (after further
processing) is included with the `gep2pep`. The excerpt must be
considered as a dummy example, as it only includes 500 genes for 5
perturbagens. It can be loaded as follows:

```{r}
geps <- readRDS(system.file("testgep.RDS", package="gep2pep"))
dim(geps)
geps[1:5, 1:3]
```

The GEPs can be converted to PEPs using the `buildPEPs` function. They
are stored as repository items by the names "category_subcategory".

```{r}
buildPEPs(rp, geps)
```

Each PEP is composed of an Enrichment Score (ES) -- p-value (PV) pair
associated to each pathway. ESs and PVs are stored in two separated
matrices. For each perturbagen, the p-value reports wether a pathway
is significantly dysregulated and the sign of the corresponding ES
indicates the direction (UP- or DOWN-regulation).

```{r}
rp$get("C3_TFT")$ES[1:3, 1:3]
rp$get("C3_TFT")$PV[1:3, 1:3]
```

## Performing Analysis

### Performing Perturbagen-Set Enrichment Analysis (PertSEA).

Suppose the stored PEPs correspond to pharmacological
perturbations. Then `gep2pep` can perform Drug-Set Enrichment Analysis
(DSEA, see Napolitano et al., 2016, Bioinformatics). It finds pathways
that are consistently dysregulated by a set of drugs, as opposed to a
background of other drugs. Of course PEPs may refer to
non-pharmacological conditions (genetic perturbations, disease states,
etc.) for analogous analyses (Perturbagen-Set Enrichment Analysis,
PertSEA). Given a set `pgset` of drugs of interest, PertSEA (which in
this case is a DSEA) is performed as follows:

```{r}
pgset <- c("(+)_chelidonine", "(+/_)_catechin")
psea <- PertSEA(rp, pgset)
```

The result is a list of of 2 elements, named "PertSEA" and "details",
the most important of which is the former. The "PertSEA" entry is a
list in which each element refers to a pathway collection and is made
of a 2-columns matrix including ESs and p-values for each pathway.

```{r}
psea$PertSEA$C3_TFT
```

In this dummy example the statistical background is made of only 3
GEPs (we added 5 in total), thus, as expected, there are no
significant p-values. For the C3_MIR collection, the pathway most
UP-regulated by the chosen set of two drugs is M5012, while the most
DOWN-regulated is M18759. They are respectively described as:

```{r}
sets <- rp$get("C3_MIR_sets")
sets$M5012$desc
sets$M18759$desc
```

The analysis can be exported in XLS format as follows:

```{r, eval=FALSE}
exportSEA(rp, psea)
```

Performing `PertSEA` using Pathway Expression Profiles derived from
drug-induced gene expression profiles yields Drug-set enrichment
analysis (DSEA). See the following reference:

* Napolitano F. et al, Drug-set enrichment analysis: a novel tool to
  investigate drug mode of action. Bioinformatics 32, 235-241 (2016).


### Performing Pathway-Set Enrichment Analysis (PathSEA)

A complementary approach to PertSEA is Pathway-Set Enrichment Analysis
(PathSEA). PathSEA searches for perturbagens that consistently
dysregulate a set of pathways. It can be seen as a pathway-based
version of the popular Gene Set Enrichment Analysis (GSEA). The
PathSEA is run independently in each pathway collection.

```{r}
pathways <- c("M11607", "M10817", "M16694",         ## from C3_TFT
              "M19723", "M5038", "M13419", "M1094") ## from C4_CGN
subdb <- db[pathways]

psea <- PathSEA(rp, subdb)
psea$PathSEA$C3_TFT
```

PathSEA results are analogous to those of PertSEA, but
perturbagen-wise. A set of pathways con also be obtained starting from
a gene of interest, for example:

```{r}
pathways <- gene2pathways(rp, "FAM126A")
```

Using a gene to obtain the pathways and performing `PathSEA` with
drug-induced Pathway Expression Profiles yields "gene2drug" analysis,
see the following reference:

* Napolitano F. et al, gene2drug: a Computational Tool for
  Pathway-based Rational Drug Repositioning, bioRxiv (2017) 192005;
  doi: https://doi.org/10.1101/192005


## Advanced access to the repository

`gep2pep` users don't need to understand how to interact with a
`gep2pep` repository, however it can be useful in some
cases. `gep2pep` repositories are in the `repo` format (see the `repo`
package), so they can be accessed as any other `repo`
repository. However item tags should not be changed, as they are used
by `gep2pep` to identify data types. Each `gep2pep` repository always
contains a special *project* item including repository and session
information, which can be shown as follows:

```{r}
rp$info("gep2pep repository")
```

Poject name and description can be provided when creating the
repository (`createRepository`), or edited with `rp$set("gep2pep
repository", newname="my project name", description="my project
description")`. The repository will also contain an item for each
pathway collection, and possibly an item for each corresponding PEP
collection, as in this example:

```{r}
rp
```

In order to look at the space that the repository is using, the
following command can be used:

```{r}
rp$info()
```


`put`, `set` and other `repo` commands can be used to alter repository
contents directly, however this could leave the repository in an
inconsistent state. The following checks a repository for possible
problems:

```{r}
checkRepository(rp)
```

```{r include=FALSE}
unlink(repoRoot, TRUE)
```


<hr/>
<small><i>
Based on `gep2pep` version `r packageVersion("gep2pep")`
</i></small>
